File: leetcode/go/sorted_squares.go
package main

import "math"

func sortedSquares(nums []int) []int {
	l := 0
	r := len(nums) - 1

	var res = make([]int, len(nums))

	for i := len(nums) - 1; i >= 0; i-- {
		if math.Abs(float64(nums[l])) > math.Abs(float64(nums[r])) {
			res[i] = nums[l] * nums[l]
			l++
		} else {
			res[i] = nums[r] * nums[r]
			r--
		}
	}

	return res
}

File: leetcode/rust/src/sorted_squares.rs
#[allow(dead_code)]
fn sorted_squares(lst: Vec<i32>) -> Vec<i32> {
    let mut left = 0;
    let mut right = lst.len() - 1;

    let mut res = vec![0; lst.len()];

    for i in (0..lst.len()).rev() {
        if i32::abs(lst[left]) > i32::abs(lst[right]) {
            res[i] = lst[left] * lst[left];

            left += 1;
        } else {
            res[i] = lst[right] * lst[right];

            right -= 1;
        }
    }

    res
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_sorted_squares_1() {
        let lst = vec![-4, -1, 0, 3, 10];
        let result = sorted_squares(lst);

        assert_eq!(result, vec![0, 1, 9, 16, 100])
    }
}
